File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/utilities/SimpleMecanumDrive.java
package org.firstinspires.ftc.teamcode.utilities;

import static com.qualcomm.robotcore.hardware.DcMotor.ZeroPowerBehavior.BRAKE;

import com.qualcomm.robotcore.hardware.DcMotor;
import com.qualcomm.robotcore.hardware.DcMotorSimple;
import com.qualcomm.robotcore.hardware.HardwareMap;

import java.util.HashMap;
import java.util.Objects;

public class SimpleMecanumDrive {

    public SimpleMecanumDrive(HardwareMap hmap) {
        fl = hmap.get(DcMotor.class, CONFIG.FRONT_LEFT);
        fr = hmap.get(DcMotor.class, CONFIG.FRONT_RIGHT);
        bl = hmap.get(DcMotor.class, CONFIG.BACK_LEFT);
        br = hmap.get(DcMotor.class, CONFIG.BACK_RIGHT);

        initMotor(fl, DcMotorSimple.Direction.FORWARD);
        initMotor(bl, DcMotorSimple.Direction.FORWARD);
        initMotor(fr, DcMotorSimple.Direction.REVERSE);
        initMotor(br, DcMotorSimple.Direction.REVERSE);
/*
        fl.setDirection(DcMotorSimple.Direction.FORWARD);
        bl.setDirection(DcMotorSimple.Direction.FORWARD);
        fr.setDirection(DcMotorSimple.Direction.REVERSE);
        br.setDirection(DcMotorSimple.Direction.REVERSE);
*/
/*
        directions.put(fl, new double[]{1, 1});
        directions.put(bl, new double[]{1, -1});
        directions.put(fr, new double[]{-1, 1});
        directions.put(br, new double[]{-1, -1});
 */
    }
    public void initMotor(
        DcMotor motor,
        DcMotorSimple.Direction direction   // delete if not awesome
//        double[] doubles                     // ^^^ same here ^^^
    )
    {
        // minimizes redundant and/or repetitive code
        motor.setZeroPowerBehavior(DcMotor.ZeroPowerBehavior.BRAKE);
        motor.setMode(DcMotor.RunMode.STOP_AND_RESET_ENCODER);
        motor.setMode(DcMotor.RunMode.RUN_WITHOUT_ENCODER);
        motor.setZeroPowerBehavior(BRAKE);

        // might work, might not work. we shall see
        motor.setDirection(direction); // idk about this. perchance it works
//        directions.put(motor, doubles);
    }
    public static DcMotor fl, fr, bl, br;

    public static HashMap<DcMotor, double[]> directions = new HashMap<>();

    public void move(double x, double y, double turn) {

        // dot of fl and br
        double dot_fl = dot(Objects.requireNonNull(directions.get(fl)), new double[]{x, y}) + turn;
        double dot_fr = dot(Objects.requireNonNull(directions.get(fr)), new double[]{x, y}) + turn;
        double dot_bl = dot(Objects.requireNonNull(directions.get(bl)), new double[]{x, y}) - turn;
        double dot_br = dot(Objects.requireNonNull(directions.get(br)), new double[]{x, y}) - turn;

        double max = Math.max(1, Math.max(Math.max(Math.abs(dot_fl), Math.abs(dot_fr)), Math.max(Math.abs(dot_bl), Math.abs(dot_br))));
        fl.setPower(dot_fl / max);
        br.setPower(dot_br / max);
        fr.setPower(dot_fr / max);
        bl.setPower(dot_bl / max);
    }
    public void tankMove(double lx, double ly, double rx, double ry) {
        fl.setPower(ly);
        bl.setPower(ly); // rotates faster than the others and is the axis the robot rotates on
        fr.setPower(ry);
        br.setPower(ry);
//        if (Math.abs(y) > 0.7) {
//            fl.setPower(y);
//        }
    }
//    public void moveFL(double x, double y, double turn) {
//        double dot_fl = dot(Objects.requireNonNull(directions.get(fl)), new double[]{x, y}) + turn;
//        double dot_fr = (dot(Objects.requireNonNull(directions.get(fr)), new double[]{x, y}) + turn);
//        double dot_bl = dot(Objects.requireNonNull(directions.get(bl)), new double[]{x, y}) - turn;
//        double dot_br = dot(Objects.requireNonNull(directions.get(br)), new double[]{x, y}) - turn;
//
//        double max = Math.max(1, Math.max(Math.max(Math.abs(dot_fl), Math.abs(dot_fr)), Math.max(Math.abs(dot_bl), Math.abs(dot_br))));
//        fl.setPower(dot_fl / max);
//        br.setPower(0);
//        fr.setPower(0);
//        bl.setPower(0);
//    }
//    public void moveFR(double x, double y, double turn) {
//        double dot_fl = dot(Objects.requireNonNull(directions.get(fl)), new double[]{x, y}) + turn;
//        double dot_fr = (dot(Objects.requireNonNull(directions.get(fr)), new double[]{x, y}) + turn);
//        double dot_bl = dot(Objects.requireNonNull(directions.get(bl)), new double[]{x, y}) - turn;
//        double dot_br = dot(Objects.requireNonNull(directions.get(br)), new double[]{x, y}) - turn;
//
//        double max = Math.max(1, Math.max(Math.max(Math.abs(dot_fl), Math.abs(dot_fr)), Math.max(Math.abs(dot_bl), Math.abs(dot_br))));
//        fl.setPower(0);
//        br.setPower(0);
//        fr.setPower(dot_fr / max);
//        bl.setPower(0);
//    }
//
//    public void moveBL(double x, double y, double turn) {
//        double dot_fl = dot(Objects.requireNonNull(directions.get(fl)), new double[]{x, y}) + turn;
//        double dot_fr = (dot(Objects.requireNonNull(directions.get(fr)), new double[]{x, y}) + turn);
//        double dot_bl = dot(Objects.requireNonNull(directions.get(bl)), new double[]{x, y}) - turn;
//        double dot_br = dot(Objects.requireNonNull(directions.get(br)), new double[]{x, y}) - turn;
//
//        double max = Math.max(1, Math.max(Math.max(Math.abs(dot_fl), Math.abs(dot_fr)), Math.max(Math.abs(dot_bl), Math.abs(dot_br))));
//        fl.setPower(0);
//        br.setPower(0);
//        fr.setPower(0);
//        bl.setPower(dot_bl / max);
//    }
//
//    public void moveBR(double x, double y, double turn) {
//        double dot_fl = dot(Objects.requireNonNull(directions.get(fl)), new double[]{x, y}) + turn;
//        double dot_fr = (dot(Objects.requireNonNull(directions.get(fr)), new double[]{x, y}) + turn);
//        double dot_bl = dot(Objects.requireNonNull(directions.get(bl)), new double[]{x, y}) - turn;
//        double dot_br = dot(Objects.requireNonNull(directions.get(br)), new double[]{x, y}) - turn;
//
//        double max = Math.max(1, Math.max(Math.max(Math.abs(dot_fl), Math.abs(dot_fr)), Math.max(Math.abs(dot_bl), Math.abs(dot_br))));
//        fl.setPower(0);
//        br.setPower(dot_br / max);
//        fr.setPower(0);
//        bl.setPower(0);
//    }

    // Each double[] will be a direction vector of length two
    public double dot(double[] a, double[] b) {
        return a[0] * b[0] + a[1] * b[1];
    }
    public int getfl() {
        return fl.getCurrentPosition();
    }
    public int getbr() {
        return br.getCurrentPosition();
    }
    public int getfr() {
        return fr.getCurrentPosition();
    }
    public int getbl() {
        return bl.getCurrentPosition();
    }

}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/utilities/SimpleMecanumDrive.kt
package org.firstinspires.ftc.teamcode.utilities

import com.qualcomm.robotcore.hardware.DcMotor
import com.qualcomm.robotcore.hardware.DcMotor.ZeroPowerBehavior
import com.qualcomm.robotcore.hardware.DcMotorSimple
import com.qualcomm.robotcore.hardware.HardwareMap
import java.util.Objects
import kotlin.math.abs
import kotlin.math.max

class SimpleMecanumDrive(hmap: HardwareMap) {
    init {
        fl = hmap.get(DcMotor::class.java, CONFIG.FRONT_LEFT)
        fr = hmap.get(DcMotor::class.java, CONFIG.FRONT_RIGHT)
        bl = hmap.get(DcMotor::class.java, CONFIG.BACK_LEFT)
        br = hmap.get(DcMotor::class.java, CONFIG.BACK_RIGHT)

        fl.zeroPowerBehavior = ZeroPowerBehavior.BRAKE
        bl.zeroPowerBehavior = ZeroPowerBehavior.BRAKE
        fr.zeroPowerBehavior = ZeroPowerBehavior.BRAKE
        br.zeroPowerBehavior = ZeroPowerBehavior.BRAKE

        fl.direction = DcMotorSimple.Direction.FORWARD
        bl.direction = DcMotorSimple.Direction.REVERSE
        fr.direction = DcMotorSimple.Direction.REVERSE // origina lly FORWARD
        br.direction = DcMotorSimple.Direction.REVERSE

        directions[fl] = doubleArrayOf(1.0, 1.0)
        directions[fr] = doubleArrayOf(1.0, -1.0)
        directions[bl] =
            doubleArrayOf(1.0, -1.0)
        directions[br] = doubleArrayOf(1.0, 1.0)
    }

    fun move(x: Double, y: Double, turn: Double) {
        // dot of fl and br

        val dot_fl = dot(Objects.requireNonNull(directions[fl]), doubleArrayOf(x, y)) - turn
        val dot_fr = (dot(
            Objects.requireNonNull(
                directions[fr]
            ), doubleArrayOf(x, y)
        ) + turn)
        val dot_bl = dot(Objects.requireNonNull(directions[bl]), doubleArrayOf(x, y)) - turn
        val dot_br = dot(Objects.requireNonNull(directions[br]), doubleArrayOf(x, y)) + turn

        val max = max(1.0, max(max(abs(dot_fl), abs(dot_fr)), max(abs(dot_bl), abs(dot_br))))
        fl.power = dot_fl / max
        br.power = dot_br / max
        fr.power = dot_fr / max
        bl.power = dot_bl / max
    }

    fun moveFL(x: Double, y: Double, turn: Double) {
        val dot_fl = dot(Objects.requireNonNull(directions[fl]), doubleArrayOf(x, y)) - turn
        val dot_fr = (dot(
            Objects.requireNonNull(
                directions[fr]
            ), doubleArrayOf(x, y)
        ) + turn)
        val dot_bl = dot(Objects.requireNonNull(directions[bl]), doubleArrayOf(x, y)) - turn
        val dot_br = dot(Objects.requireNonNull(directions[br]), doubleArrayOf(x, y)) + turn

        val max = max(1.0, max(max(abs(dot_fl), abs(dot_fr)), max(abs(dot_bl), abs(dot_br))))
        fl.power = dot_fl / max
        br.power = 0.0
        fr.power = 0.0
        bl.power = 0.0
    }

    fun moveFR(x: Double, y: Double, turn: Double) {
        val dot_fl = dot(Objects.requireNonNull(directions[fl]), doubleArrayOf(x, y)) - turn
        val dot_fr = (dot(
            Objects.requireNonNull(
                directions[fr]
            ), doubleArrayOf(x, y)
        ) + turn)
        val dot_bl = dot(Objects.requireNonNull(directions[bl]), doubleArrayOf(x, y)) - turn
        val dot_br = dot(Objects.requireNonNull(directions[br]), doubleArrayOf(x, y)) + turn

        val max = max(1.0, max(max(abs(dot_fl), abs(dot_fr)), max(abs(dot_bl), abs(dot_br))))
        fl.power = 0.0
        br.power = 0.0
        fr.power = dot_fr / max
        bl.power = 0.0
    }

    fun moveBL(x: Double, y: Double, turn: Double) {
        val dot_fl = dot(Objects.requireNonNull(directions[fl]), doubleArrayOf(x, y)) - turn
        val dot_fr = (dot(
            Objects.requireNonNull(
                directions[fr]
            ), doubleArrayOf(x, y)
        ) + turn)
        val dot_bl = dot(Objects.requireNonNull(directions[bl]), doubleArrayOf(x, y)) - turn
        val dot_br = dot(Objects.requireNonNull(directions[br]), doubleArrayOf(x, y)) + turn

        val max = max(1.0, max(max(abs(dot_fl), abs(dot_fr)), max(abs(dot_bl), abs(dot_br))))
        fl.power = 0.0
        br.power = 0.0
        fr.power = 0.0
        bl.power = dot_bl / max
    }

    fun moveBR(x: Double, y: Double, turn: Double) {
        val dot_fl = dot(Objects.requireNonNull(directions[fl]), doubleArrayOf(x, y)) - turn
        val dot_fr = (dot(
            Objects.requireNonNull(
                directions[fr]
            ), doubleArrayOf(x, y)
        ) + turn)
        val dot_bl = dot(Objects.requireNonNull(directions[bl]), doubleArrayOf(x, y)) - turn
        val dot_br = dot(Objects.requireNonNull(directions[br]), doubleArrayOf(x, y)) + turn

        val max = max(1.0, max(max(abs(dot_fl), abs(dot_fr)), max(abs(dot_bl), abs(dot_br))))
        fl.power = 0.0
        br.power = dot_br / max
        fr.power = 0.0
        bl.power = 0.0
    }

    // Each double[] will be a direction vector of length two
    fun dot(a: DoubleArray, b: DoubleArray): Double {
        return a[0] * b[0] + a[1] * b[1]
    }


    companion object {
        var fl: DcMotor
        var fr: DcMotor
        var bl: DcMotor
        var br: DcMotor

        var directions: HashMap<DcMotor, DoubleArray> = HashMap()
    }
}